# Statistical Data Analysis 

In this section of the project, we will discover the Pearson’s chi-squared statistical hypothesis test for quantifying the independence of pairs of categorical variables.

In the case of classification problems where input variables are also categorical, we can use statistical tests to determine whether the output variable is dependent or independent of the input variables. If independent, then the input variable is a candidate for a feature that may be irrelevant to the problem and removed from the dataset.

Therefore, our hypothesis statements are going to be the following

    H0: X and Y are independent
    H1: X and Y are dependent

Here, we explore if the features Gender and Nationality are dependent to the target variable i.e. if the user pays for apps.

In [1]:
import pandas as pd
import scipy.stats as stats
pd.options.mode.chained_assignment = None

In [2]:
clean = pd.read_excel('clean_data.xlsx', index_col=0)

In [3]:
complete = clean[clean.response == 1]

In [4]:
complete.head()

startdate             enddate  response  participant_type  \
id                                                                       
2  2012-09-26 07:45:19 2012-09-26 07:56:41         1                 1   
3  2012-09-26 07:45:35 2012-09-26 08:01:56         1                 1   
4  2012-09-26 16:58:29 2012-09-26 17:05:50         1                 1   
5  2012-09-27 04:16:04 2012-09-27 04:24:56         1                 1   
6  2012-09-27 08:50:34 2012-09-27 08:56:48         1                 1   

    browser browser_version                               os  \
id                                                             
2    Safari               6  CPU iPhone OS 6_0 like Mac OS X   
3    Safari               6         CPU OS 6_0 like Mac OS X   
4   Firefox          15.0.1              Intel Mac OS X 10.6   
5    Chrome    22.0.1229.79            Intel Mac OS X 10_7_4   
6    Chrome    21.0.1180.89                            WOW64   

   screen_resolution flash_version  java_support  ...  \
id                                                ...   
2            320x480            -1           0.0  ...   
3           768x1024            -1           0.0  ...   
4          1920x1200      11.4.402           1.0  ...   
5           1280x800      11.4.402           1.0  ...   
6          1920x1080       11.3.31           1.0  ...   

   extraverted_enthusiastic  critical_quarrelsome dependable_self_disciplined  \
id                                                                              
2                       6.0                   3.0                         7.0   
3                       4.0                   4.0                         5.0   
4                       4.0                   3.0                         6.0   
5                       2.0                   6.0                         4.0   
6                       3.0                   2.0                         6.0   

   anxious_easily_upset  open_to_new_experiences_complex reserved_quiet  \
id                                                                        
2                   2.0                              6.0            3.0   
3                   2.0                              3.0            3.0   
4                   3.0                              5.0            5.0   
5                   3.0                              6.0            5.0   
6                   2.0                              6.0            4.0   

   sympathetic_warm  disorganized_careless  calm_emotionally_stable  \
id                                                                    
2               4.0                    3.0                      4.0   
3               5.0                    3.0                      5.0   
4               5.0                    2.0                      5.0   
5               7.0                    3.0                      5.0   
6               3.0                    2.0                      5.0   

    conventional_uncreative  
id                           
2                       4.0  
3                       3.0  
4                       3.0  
5                       3.0  
6                       2.0  

[5 rows x 160 columns]

In [5]:
complete.do_not_pay_for_apps = complete.do_not_pay_for_apps.fillna(0)

In [6]:
complete.do_not_pay_for_apps

id
2        1.0
3        1.0
4        1.0
5        0.0
6        1.0
7        0.0
8        1.0
9        0.0
11       1.0
13       0.0
16       0.0
17       1.0
18       0.0
19       1.0
20       1.0
21       1.0
22       0.0
23       1.0
25       1.0
26       1.0
27       1.0
28       1.0
33       0.0
34       0.0
36       1.0
38       1.0
39       1.0
40       1.0
41       0.0
42       0.0
        ... 
9933     1.0
9934     0.0
9936     0.0
9939     1.0
9942     1.0
9943     1.0
9944     0.0
9945     1.0
9959     1.0
9966     1.0
9971     1.0
9981     0.0
10007    0.0
10010    1.0
10017    1.0
10027    1.0
10030    1.0
10034    1.0
10041    1.0
10046    1.0
10056    0.0
10060    1.0
10069    1.0
10071    1.0
10073    1.0
10086    1.0
10092    1.0
10111    1.0
10112    1.0
10208    1.0
Name: do_not_pay_for_apps, Length: 4824, dtype: float64

**Gender**

    H0: Paying for apps is independent of the gender of the users
    H1: Paying for apps is dependent with the gender of the users

In [7]:
gender_pays_df = complete[['gender', 'do_not_pay_for_apps']]
gender_pays_df

gender  do_not_pay_for_apps
id                                
2        Male                  1.0
3        Male                  1.0
4        Male                  1.0
5        Male                  0.0
6      Female                  1.0
7        Male                  0.0
8        Male                  1.0
9        Male                  0.0
11     Female                  1.0
13     Female                  0.0
16       Male                  0.0
17       Male                  1.0
18       Male                  0.0
19     Female                  1.0
20     Female                  1.0
21     Female                  1.0
22       Male                  0.0
23       Male                  1.0
25       Male                  1.0
26       Male                  1.0
27     Female                  1.0
28       Male                  1.0
33       Male                  0.0
34     Female                  0.0
36       Male                  1.0
38       Male                  1.0
39       Male                  1.0
40       Male                  1.0
41       Male                  0.0
42       Male                  0.0
...       ...                  ...
9933   Female                  1.0
9934     Male                  0.0
9936     Male                  0.0
9939   Female                  1.0
9942     Male                  1.0
9943   Female                  1.0
9944     Male                  0.0
9945     Male                  1.0
9959   Female                  1.0
9966   Female                  1.0
9971     Male                  1.0
9981   Female                  0.0
10007    Male                  0.0
10010    Male                  1.0
10017    Male                  1.0
10027    Male                  1.0
10030    Male                  1.0
10034    Male                  1.0
10041    Male                  1.0
10046  Female                  1.0
10056  Female                  0.0
10060  Female                  1.0
10069    Male                  1.0
10071  Female                  1.0
10073  Female                  1.0
10086    Male                  1.0
10092    Male                  1.0
10111  Female                  1.0
10112  Female                  1.0
10208  Female                  1.0

[4824 rows x 2 columns]

In [8]:
gender_pays = gender_pays_df.groupby('gender').sum()
gender_pays

do_not_pay_for_apps
gender                     
Female               1518.0
Male                 1209.0

In [9]:
gender_pays['pay_for_apps'] = gender_pays_df.groupby('gender').count() - gender_pays_df.groupby('gender').sum()
gender_pays

do_not_pay_for_apps  pay_for_apps
gender                                   
Female               1518.0         960.0
Male                 1209.0        1137.0

In [10]:
stats.chi2_contingency(gender_pays)

(45.981011507353415,
 1.1940473451340507e-11,
 1,
 array([[1400.80970149, 1077.19029851],
        [1326.19029851, 1019.80970149]]))

For degree of freedom 1, we have a very high chi-squared value of 45.98 well above the table value of 3.84 so we have enough statistical to conclude that the two variable are dependent.

**Country**

    H0: Paying for apps is independent of the nationality of the users
    H1: Paying for apps is dependent with the nationality of the users

In [11]:
country_pays_df = complete[['nationality', 'do_not_pay_for_apps']]
country_pays_df

nationality  do_not_pay_for_apps
id                                    
2             USA                  1.0
3           Other                  1.0
4           Italy                  1.0
5          Mexico                  0.0
6              UK                  1.0
7           Other                  0.0
8       Australia                  1.0
9           Japan                  0.0
11      Australia                  1.0
13          Japan                  0.0
16          Japan                  0.0
17          Other                  1.0
18          China                  0.0
19          Other                  1.0
20             UK                  1.0
21          Other                  1.0
22          Italy                  0.0
23         France                  1.0
25          Italy                  1.0
26          Italy                  1.0
27          Other                  1.0
28          Italy                  1.0
33          Italy                  0.0
34         Mexico                  0.0
36          Italy                  1.0
38          Italy                  1.0
39         France                  1.0
40          Italy                  1.0
41          Other                  0.0
42          Italy                  0.0
...           ...                  ...
9933        Japan                  1.0
9934        Japan                  0.0
9936        Japan                  0.0
9939        Japan                  1.0
9942        Japan                  1.0
9943        Japan                  1.0
9944        Japan                  0.0
9945          USA                  1.0
9959        Japan                  1.0
9966        Japan                  1.0
9971        Japan                  1.0
9981        Japan                  0.0
10007       Japan                  0.0
10010       Japan                  1.0
10017       Japan                  1.0
10027       Japan                  1.0
10030       Japan                  1.0
10034       Japan                  1.0
10041       Japan                  1.0
10046       Japan                  1.0
10056       Japan                  0.0
10060       Japan                  1.0
10069       Japan                  1.0
10071       Japan                  1.0
10073       Japan                  1.0
10086       Japan                  1.0
10092       Japan                  1.0
10111         USA                  1.0
10112       Japan                  1.0
10208       Japan                  1.0

[4824 rows x 2 columns]

In [12]:
country_pays = country_pays_df.groupby('nationality').sum()
country_pays

do_not_pay_for_apps
nationality                     
Australia                  147.0
Brazil                     252.0
Canada                     246.0
China                      162.0
France                     124.0
Germany                    150.0
India                      204.0
Italy                      166.0
Japan                      160.0
Korea                      203.0
Mexico                      95.0
Other                      166.0
Russia                     160.0
Spain                      155.0
UK                         155.0
USA                        182.0

In [13]:
country_pays['pay_for_apps'] = country_pays_df.groupby('nationality').count() - country_pays_df.groupby('nationality').sum()
country_pays

do_not_pay_for_apps  pay_for_apps
nationality                                   
Australia                  147.0         100.0
Brazil                     252.0         176.0
Canada                     246.0         180.0
China                      162.0         395.0
France                     124.0         109.0
Germany                    150.0         100.0
India                      204.0         161.0
Italy                      166.0         100.0
Japan                      160.0          54.0
Korea                      203.0          61.0
Mexico                      95.0         150.0
Other                      166.0          95.0
Russia                     160.0         127.0
Spain                      155.0          96.0
UK                         155.0          95.0
USA                        182.0          98.0

In [14]:
stats.chi2_contingency(country_pays)

(303.2415112441552,
 1.1847806153179469e-55,
 15,
 array([[139.62873134, 107.37126866],
        [241.94776119, 186.05223881],
        [240.81716418, 185.18283582],
        [314.87126866, 242.12873134],
        [131.71455224, 101.28544776],
        [141.32462687, 108.67537313],
        [206.33395522, 158.66604478],
        [150.36940299, 115.63059701],
        [120.9738806 ,  93.0261194 ],
        [149.23880597, 114.76119403],
        [138.49813433, 106.50186567],
        [147.54291045, 113.45708955],
        [162.24067164, 124.75932836],
        [141.88992537, 109.11007463],
        [141.32462687, 108.67537313],
        [158.28358209, 121.71641791]]))

For degree of freedom 1, we have a very high chi-squared value of 45.98 well above the table value of 3.84 so we have enough statistical to conclude that the two variable are dependent.